control-plane/typeSpecConverter.cpp

```diff
@@ -41,33 +41,49 @@ namespace P4 {
 
 namespace ControlPlaneAPI {
 
-bool hasTranslationAnnotation(const IR::Type* type, std::string* uri, int* sdnB) {
+bool hasTranslationAnnotation(const IR::Type* type,
+                              TranslationAnnotation* payload) {
     auto ann = type->getAnnotation("p4runtime_translation");
     if (!ann) return false;
-    auto uriL = ann->expr[0]->to<IR::StringLiteral>();
-    if (!uriL) {
-        ::error(ErrorType::ERR_INVALID,
-                "%1%: the first argument to @p4runtime_translation must be a string literal",
-                type);
-        return false;
-    }
-    *uri = static_cast<std::string>(uriL->value);
-    auto sdnBL = ann->expr[1]->to<IR::Constant>();
-    if (!sdnBL) {
-        ::error(ErrorType::ERR_INVALID,
-                "%1%: the second argument to @p4runtime_translation must be a positive constant",
-                type);
-        return false;
-    }
-    if (sdnBL->value <= 0 || sdnBL->value > std::numeric_limits<int32_t>::max()) {
-        ::error(ErrorType::ERR_INVALID,
-                "%1%: the second argument to @p4runtime_translation must be positive integer that "
-                "fits in 32 bits",
-                type);
-        return false;
+
+    // Syntax: @pruntime_translation(<uri>, <basic_type>).
+    BUG_CHECK(ann->expr.size() == 2,
+              "%1%: expected @p4runtime_translation annotation with 2 "
+              "arguments, but found %2% arguments",
+              type, ann->expr.size());
+    const IR::Expression* first_arg = ann->expr[0];
+    const IR::Expression* second_arg = ann->expr[1];
+
+    auto uri = first_arg->to<IR::StringLiteral>();
+    BUG_CHECK(uri != nullptr,
+              "%1%: the first argument to @p4runtime_translation must be a "
+              "string literal",
+              type);
+    payload->original_type_uri = std::string(uri->value);
+
+    BUG_CHECK(second_arg->to<IR::DefaultExpression>() != nullptr,
```

Comment:
@smolkaj @mbudiu-vmw this doesn't seem right. Isn't a `DefaultExpression` basically a "don't care" expression? I think you meant to use `Expression` here but I am not very familiar with the IR in general.

Reply:
You're right, a DefaultExpression is a "don't care". I really want an IR::Type here, however the parser must return a vector of `IR::Expression`s for things to type check. In other words, using a `DefaultExpression` (that carries the type I'm interested in) is a hack;  I cannot use `Expression` directly since that is an abstract class.